src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs

```diff
@@ -33,30 +39,38 @@ public QuicConnectionListener(QuicTransportOptions options, ILogger log, EndPoin
             throw new InvalidOperationException($"QUIC doesn't support listening on the configured endpoint type. Expected {nameof(IPEndPoint)} but got {endpoint.GetType().Name}.");
         }
 
-        if (sslServerAuthenticationOptions.ApplicationProtocols is null || sslServerAuthenticationOptions.ApplicationProtocols.Count == 0)
+        if (protocols.Count == 0)
         {
             throw new InvalidOperationException("No application protocols specified.");
         }
 
         _log = log;
+        _protocols = protocols;
         _context = new QuicTransportContext(_log, options);
         _quicListenerOptions = new QuicListenerOptions
         {
-            ApplicationProtocols = sslServerAuthenticationOptions.ApplicationProtocols,
+            ApplicationProtocols = _protocols,
             ListenEndPoint = listenEndPoint,
             ListenBacklog = options.Backlog,
-            ConnectionOptionsCallback = (connection, helloInfo, cancellationToken) =>
+            ConnectionOptionsCallback = async (connection, helloInfo, cancellationToken) =>
             {
+                var serverAuthenticationOptions = await sslServerAuthenticationOptionsCallback(helloInfo, cancellationToken);
+
+                // If the SslServerAuthenticationOptions doesn't have a cert or protocols then the
+                // QUIC connection will fail and the client receives an unhelpful message.
+                // Validate the options on the server and log issues to improve debugging.
+                ValidateServerAuthenticationOptions(serverAuthenticationOptions);
```

Comment:
In another code path we set APLN for them if it wasn't set in the callback:
https://github.com/dotnet/aspnetcore/blob/5834214ddeab320106c8e728b92b611a8c6add63/src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs#L431-L437